src/snap-keyring.ts

```diff
@@ -0,0 +1,355 @@
+
+  #addressToSnapId: Record<string, string>;
+
+  #pendingRequests: Record<string, DeferredPromise<any>>;
+
+  constructor(controller: SnapController) {
+    super();
+    this.type = SnapKeyring.type;
+    this.#snapClient = new KeyringSnapControllerClient({ controller });
+    this.#addressToAccount = {};
+    this.#addressToSnapId = {};
+    this.#pendingRequests = {};
+  }
+
+  /**
+   * Sync accounts from all snaps.
+   *
+   * @param extraSnapIds - List of extra snap IDs to include in the sync.
+   */
+  async #syncAccounts(...extraSnapIds: string[]): Promise<void> {
+    // Add new snap IDs to the list.
+    const snapIds = Object.values(this.#addressToSnapId).concat(extraSnapIds);
+
+    // Remove all addresses from the maps.
+    this.#addressToAccount = {};
+    this.#addressToSnapId = {};
+
+    // ... And add them back.
+    for (const snapId of unique(snapIds)) {
+      const accounts = await this.#snapClient.withSnapId(snapId).listAccounts();
+      for (const account of accounts) {
+        this.#addressToAccount[account.address] = account;
+        this.#addressToSnapId[account.address] = snapId;
+      }
+    }
+  }
+
+  /**
+   * Handle a message from a snap.
+   *
+   * @param snapId - ID of the snap.
+   * @param message - Message sent by the snap.
+   * @param saveSnapKeyring - Function to save the snap's state.
+   * @returns The execution result.
+   */
+  async handleKeyringSnapMessage(
+    snapId: string,
+    message: unknown,
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    saveSnapKeyring: Function,
+  ): Promise<Json> {
+    assert(message, SnapMessageStruct);
+    const [method, params] = message;
+    switch (method) {
+      case 'update':
+      case 'delete':
+      case 'create': {
+        await this.#syncAccounts(snapId);
+        await saveSnapKeyring();
+        return null;
+      }
+
+      case 'read': {
+        return await this.#listAccounts(snapId);
+      }
+
+      case 'submit': {
+        const { id, result } = params as any; // FIXME: add a struct for this
+        this.#resolveRequest(id, result);
+        return true;
+      }
+
+      default:
+        throw ethErrors.rpc.invalidParams({
+          message: 'Must specify a valid snap_manageAccounts "methodName".',
+        });
+    }
+  }
+
+  /**
+   * Serialize the keyring state.
+   *
+   * @returns Serialized keyring state.
+   */
+  async serialize(): Promise<KeyringState> {
+    return {
+      addressToAccount: this.#addressToAccount,
+      addressToSnapId: this.#addressToSnapId,
+    };
+  }
+
+  /**
+   * Deserialize the keyring state into this keyring.
+   *
+   * @param state - Serialized keyring state.
+   */
+  async deserialize(state: KeyringState | undefined): Promise<void> {
+    if (state === undefined) {
+      return;
+    }
+
+    assert(state, KeyringStateStruct);
+    this.#addressToAccount = state.addressToAccount;
+    this.#addressToSnapId = state.addressToSnapId;
```

Comment:
we should set #pendingRequests as well

Reply:
Good point, but it may not be possible due to the promises. Maybe we can leave it for a different PR?